cmd: document target granularity for push/pull/etc, et al. (#1384)
iterative/dvc.org: content/docs/command-reference/status.md

either showing which files or directories have changed in the
<abbr>workspace</abbr> and must be added or reproduced again (with `dvc add` or
`dvc repro`); or differences between <abbr>cache</abbr> vs. remote storage
(implying `dvc push` or `dvc pull` should be run to synchronize them). The
_remote_ mode is triggered by using the `--cloud` or `--remote` options:

| Mode   | Option     | Description                                                                                                                 |
| ------ | ---------- | --------------------------------------------------------------------------------------------------------------------------- |
| local  | _none_     | Comparisons are made between data files in the workspace and corresponding files in the cache directory (e.g. `.dvc/cache`) |
| remote | `--remote` | Comparisons are made between the cache, and the given remote. Remote storage is defined using the `dvc remote` command.     |
| remote | `--cloud`  | Comparisons are made between the cache, and the default remote (typically defined with `dvc remote --default`).             |

Without arguments, this command checks all stages (in `dvc.yaml`) and `.dvc`

> files to compare the hash values of their <abbr>outputs</abbr> (in `dvc.lock`)

Reviewer: (for stages output hashes as defined in `dvc.lock`)?

Reviewer 2: "checks stages in `dvc.yaml` for stage output hash values as defined in `dvc.lock`" ? That's contradictory. Let me think 🧠

Reviewer 2: Changed to 

"checks all stages (defined in `dvc.yaml`) and `.dvc` files, and compares the hash values of their <abbr>outputs</abbr> (found in `dvc.lock` for stages) against"

in acf0196f.

Reviewer 2: #1663 can probably simplify this too.

`dvc status` searches for changes in the existing tracked data and pipelines,
either showing which files or directories have changed in the
<abbr>workspace</abbr> and must be added or reproduced again (with `dvc add` or
`dvc repro`); or differences between <abbr>cache</abbr> vs. remote storage
(implying `dvc push` or `dvc pull` should be run to synchronize them). The
_remote_ mode is triggered by using the `--cloud` or `--remote` options:

| Mode   | Option     | Description                                                                                                                 |
| ------ | ---------- | --------------------------------------------------------------------------------------------------------------------------- |
| local  | _none_     | Comparisons are made between data files in the workspace and corresponding files in the cache directory (e.g. `.dvc/cache`) |
| remote | `--remote` | Comparisons are made between the cache, and the given remote. Remote storage is defined using the `dvc remote` command.     |
| remote | `--cloud`  | Comparisons are made between the cache, and the default remote (typically defined with `dvc remote --default`).             |

> Without arguments, this command checks all stages (in `dvc.yaml` and `dvc.lock`)

Reviewer: is it union of two files? or only one of them determines the list of stages to check?

Reviewer 2: I think it's effectively the union, from a conversation I had with @efiop once: It checks dvc.yaml to determine the stages, but later those not found in dvc.lock are skipped. Maybe a warning is printed.

Author: @jorgeorpinel It is not the union, we've discussed that `dvc.yaml` is the ultimate source of all truth.

Reviewer 2: OK. But since stages not found in dvc.lock are then skipped, that's why I say it's effectively the union.

But in any case, that wasn't the intention of this text. It mentions dvc.lock because the paragraph talks about the hash values of the stage outputs. Updated a bit in 6ac8b0fc.

Reviewer: > But since stages not found in dvc.lock are then skipped

if they are skipped w/o a single message then it's intersection? Only those that exist in both are checked?

Reviewer 2: AFAIR there is a WARNING message.

and `.dvc` files to compare the hash values of their <abbr>outputs</abbr>
against the actual data files or directories in the workspace (the
`--all-branches` and `--all-tags` options enable using multiple workspace
versions).

> The `targets` given to this command (if any) limit what to check. It accepts

Reviewer: limits?

Reviewer 2: The targets (plural) limit 🙂

Reviewer: then "It accepts" sounds a bit strange I would say

Reviewer 2: True! Updated in cabc8bfd.

differences, the changes in <abbr>dependencies</abbr> and/or
<abbr>outputs</abbr> that differ are listed. For each item listed, either the
file name or hash is shown, along with a _state description_, as detailed below:

- _changed checksum_ means that the `.dvc` file hash has changed (e.g. someone
  manually edited it).

- _always changed_ means that this is a `.dvc` file with no dependencies (see
  `dvc add`) or that the stage in `dvc.yaml` has the `always_changed: true`
  value set (see `--always-changed` option in `dvc run`).

- _changed deps_ or _changed outs_ means that there are changes in dependencies
  or outputs tracked by the stage or `.dvc` file. Depending on the use case,
  commands like `dvc commit`, `dvc repro`, or `dvc run` can be used to update
  the file. Possible states are:

  - _new_: An <abbr>output</abbr> is found in the <abbr>workspace</abbr>, but
    there is no corresponding file hash saved in the `dvc.lock` or `.dvc` file
    yet.
  - _modified_: An output or <abbr>dependency</abbr> is found in the workspace,
    but the corresponding file hash in the `dvc.lock` or `.dvc` file is not up
    to date.

>   - _deleted_: The output or dependency is referenced in a `dvc.lock` or `.dvc`

Reviewer: what happens if there is not `dvc.lock`?

Reviewer 2: All data is reported as _modified_.

Reviewer: 👍

When this is combined with the `--with-deps` option, a search is made for
changes in other stages that affect each target.

> In the remote mode, if `targets` are files/directories tracked by DVC, DVC will
> only compare the cache for them, skipping the rest.

Reviewer 2: ```suggestion
In the remote mode, if `targets` are files/directories tracked by DVC, DVC will
only compare those targets, skipping the rest.
```

Reviewer 2: Solved in 94a86c5d

file name or hash is shown, and additionally a status word is shown describing
the changes (described below).

- _changed checksum_ means that the stage (in `dvc.yaml`) or `.dvc` file hash
  has changed (e.g. someone manually edited the file).

> 

Reviewer 2: Another quick one @efiop, does this status still exist? I removed it for now.

> Unrelated to the PR's original intention I know, sorry.

Author: It does still exist :)

Reviewer 2: But the file hashes of .dvc files and dvc.yaml are no longer saved so... what does this check? Maybe i'm confused.

Author: @jorgeorpinel It is is still in imported dvc-files. And `dvc add`ed onces will still react to changes in it (backward compatibility). In `dvc.yaml` we do indeed call it differently: "changed command". So I guess we could remove "in dvc.yaml" part here. But I have to point out that we should probably not extend the scope of this PR, it will be painful :slightly_smiling_face:

Reviewer 2: Yeah I will extract this. For now I just removed dvc.yaml from here. Thanks!

Reviewer 2: And extracted to iterative/dvc/issues/4189

<abbr>output</abbr> `bar` have changed. Likewise for `foo.dvc`, the dependency
`foo` has changed, but no output has changed.

## Example: Specific files or directories

`dvc status` only checks the tracked data corresponding to any given `targets`:

```dvc
$ dvc status foo.dvc dobar
foo.dvc
  changed outs:
          deleted:      foo
  changed checksum
dobar
  changed deps:
          modified:      bar
  changed outs:
          not in cache:      foo
```

> In this case, the target `foo.dvc` is a `.dvc` file to track the `foo` file,
> while `dobar` is the name of a stage defined in `dvc.yaml`.

> Note that you can check data within directories tracked as a whole, such as the

Reviewer: I wonder if we can just use "tracked directories"? omit "as a whole"

Reviewer 2: OK, removed phrase "as a whole" completely in 4cac81fa.

            modified:  code/featurization.py
```

If the `dvc status` command is limited to a target that had no changes, but by

> adding `--with-deps`, any upstream change will be found (in a preceding stage).

Reviewer: upstream can be confusing with Git terminology

Reviewer 2: OK. Updated in 8e7f63a7.

```dvc
$ dvc status --remote storage
new:      data/model.p
new:      data/eval.txt
new:      data/matrix-train.p
new:      data/matrix-test.p
```

The output shows where the location of the remote storage is, as well as any
differences between the <abbr>cache</abbr> and `storage` remote.

> ## Example: Import stage needs update

Reviewer: Imported data?

Reviewer 2: Sure, applying:
```suggestion
## Example: Imported data needs update
```

```dvc
$ dvc status --remote storage
new:      data/model.p
new:      data/eval.txt
new:      data/matrix-train.p
new:      data/matrix-test.p
```

The output shows where the location of the remote storage is, as well as any
differences between the <abbr>cache</abbr> and `storage` remote.

> ## Example: Imported data needs update

Reviewer: Check imported files ?

Reviewer 2: f280bdf1
